sys: add credman (D)TLS credential management module

pokgak opened the pull request:
### Contribution description
This PR is the first part to add sock_dtls module for (D)TLS support in RIOT. credman can also be used
to handle credentials for TLS if TLS is supported in RIOT in the future.

To use DTLS, users must first tell the DTLS sock which credential needs to be used for the
encryption and decryption process. There are two use cases here, 1) the user has direct access to the
DTLS sock and can actually give the pointer of the credential buffer to the sock, 2) the user does not have access to the sock because it is located in a lower leveled library used by user application (e.g. gcoap + DTLS). 

credman handles both of these by providing a unified way to pass credential to the sock. Even though this adds few more steps to add a credential to a sock (add credential, register tag to use), it avoids having multiple ways to tell sock which credential to use depending on the use case, which could be confusing to a new user.

As there can be multiple credentials of the same type added into the system, the (tag, type) -> credential buffer mapping is used by credman to identify which credential to use with which sock.

### Testing procedure
Added unit tests under `tests/unittests/tests-credman`.

### Issues/PRs references
See also #10897

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,180 @@
+/**
+ * @brief Tag of the credential. Must be bigger than 0.
+ */
+typedef unsigned credman_tag_t;
```
Do we really need up to 32-bits to identify a credential? I think `uint16_t` suffices here (and since you nicely abstracted the type we still go up to `uint32_t` in case I was wrong ;-)).

jcarrano replied:
Beware, that will only save memory if `-fshort-enum` is on (because of credman_type_t).

pokgak replied:
Changed to uint16_t.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,147 @@
+    /* unions only point to one address at a time,
+     * so it doesn't matter which entry in the union
+     * we copied here, as all of them are the same */
+    c->params.psk = credential->params.psk;
```
```C
    credentials[pos] = *credential
```

should also work here and might save some ROM (not sure though). It definitely is more future stable, in case `credman_credential_t` is extended.

pokgak replied:
Fixed.

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,180 @@
+    CREDMAN_EXIST       = -1,
+    CREDMAN_NO_SPACE    = -2,
+    CREDMAN_NOT_FOUND   = -3,
+    CREDMAN_ERROR       = -4,
```
Doc missing.

pokgak replied:
Docs added.

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,180 @@
+ * @return CREDMAN_OK on success
+ * @return CREDMAN_EXIST if credential of @p tag and @p type already exist
+ * @return CREDMAN_NO_SPACE if system buffer full
+ * @return CREDMAN_ERROR on other errors
```
IMHO what you use for `CREDMAN_ERROR` in the implementation are all input-parameter errors, so why not have a `CREDMAN_INVALID` as well to cover this case more clearly?

pokgak replied:
You're right. Added `CREDMAN_INVALID`.

miri64 replied:
> This is just the type definition, so adding value restrictions to it seems off (also its "[…] greater than 0").

Rather put the restriction in the (missing) doc for `CREDMAN_INVALID` (e.g. `@return CREDMAN_INVALID when […] or credman_credential_t::tag of @p credential is @ref CREDMAN_TAG_EMPTY`).

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,147 @@
+        }
+        break;
+    default:
+        return CREDMAN_ERROR;
```
Here you could return e.g. `CREDMAN_TYPE_UNKNOWN`.

pokgak replied:
Fixed.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,147 @@
+
+void credman_init(void)
+{
+    memset(credentials, 0, sizeof(credman_credential_t) * CREDMAN_MAX_CREDENTIALS);
```
This only should be required for testing. How about calling the function `credman_reset()` instead and only making it available `#ifdef TEST_SUITES`?

pokgak replied:
Just found out that all global and static variables are automatically initialized to zero. So this function is only used for testing. Changed to `credman_reset()`.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,147 @@
+#include "debug.h"
+
+static credman_credential_t credentials[CREDMAN_MAX_CREDENTIALS];
+static unsigned used = 0;
```
This variable and how you access `credentials` makes the library not thread-safe. Either document that or deal with it.

pokgak replied:
Added commit 4bdc0f0 to use mutex to make it thread-safe.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,147 @@
+
+static int _find_credential_pos(credman_tag_t tag, credman_type_t type)
+{
+    for (unsigned i = 0; i < used; i++) {
```
This does not work.

If one calls e.g.

```C
credman_add_credential(&credential1);
credman_add_credential(&credential2);
credman_add_credential(&credential3);
credman_delete_credential(&credential2);
```

`used` will be 2, so `credential3` will not be reached as its index is 2.

pokgak replied:
Fixed. Added test case to test this.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,147 @@
+    if (pos >= 0) {
+        DEBUG("credman: credential with tag %d and type %d already exist\n",
+              credential->tag, credential->type);
+        return CREDMAN_EXIST;
```
What if `credential->params` is identical to `credentials[pos]`

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,160 @@
+
+    ret = credman_add_credential(&exp_credential);
+    TEST_ASSERT_EQUAL_INT(CREDMAN_OK, ret);
+
```
A test case for 

```C
credman_add_credential(&credential1);
credman_add_credential(&credential2);
credman_add_credential(&credential3);
credman_delete_credential(&credential2);
credman_get_credential(&out_credential, credential3.tag, credential3.type);
```

should exist ;-)

pokgak replied:
Added test case.

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,160 @@
+    /* delete a deleted credential */
+    ret = credman_delete_credential(exp_credential.tag, exp_credential.type);
+    TEST_ASSERT_EQUAL_INT(CREDMAN_NOT_FOUND, ret);
+    TEST_ASSERT_EQUAL_INT(exp_count, credman_get_used_count());
```
A test case for 

```C
credman_add_credential(&credential1);
credman_add_credential(&credential2);
credman_add_credential(&credential3);
credman_delete_credential(&credential1);
credman_delete_credential(&credential2);
credman_delete_credential(&credential3);
```

should exist ;-)

pokgak replied:
Added new test case.

jcarrano commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,169 @@
+        goto end;
+    }
+
+    credentials[pos] = *credential;
```
In the description you say 

> It just holds the pointers to the credentials given by the user.

but here the value is being copied. So that the user must keep is the "*_params_t" structure _and_ the buffers pointed to by that structure.

pokgak replied:
Thanks. I missed that. `credman_credential_t` initially have `*_params_t` itself in it, not just the pointer to it. I'll need to change that back.

pokgak commented:
> I think this may be over-specialized. If the user must remember a (32 bit) numerical tag and this does not copy the buffers, how is this different from just remembering a pointer to the buffer.

The main reason for this module is so that the user can supply their own credentials for libraries that uses(D)TLS (e.g. gcoap + sock_dtls [1]). With this module, users can tell the sock_dtls used in gcoap, which credential to use without having access to the sock itself. Ideally, if the user is using sock_dtls directly, they actually doesn't need this module. It is more towards other modules/libraries that uses DTLS.

> Also, why mix the storage and validation logic. If what is needed is an int->pointer mapping (thread safe and all) why not make that a module and implement the validation and retrieval on top of it.

The (tag, type) mapping is to allow use of different credentials with different sock_dtls. If this module are intended for more general usage, mixing the storage and validation logic may not be good here but for (D)TLS it is easier to check it once when added if the credential is valid than to check it later every time the credential will be used in the (D)TLS libraries. Also, I don't know if such module (int->pointer mapping) is needed currently other than here.

> The design of this does not allow for persisting the credentials in any way.

Copying the buffer into the system might use a larger amount of memory than needed e.g. credman reserves memory enough for ECDSA (longer) but the user only uses PSK (much shorter). That is why I decided to only save pointer to the buffer.

> If credentials are known at compile time, why not use constfs. And if they are dynamically added in run-time, where will they be allocated?

I'm not aware of constfs before this. Thanks for pointing that out. I'll look into it. For the dynamically added in run-time, the users have to supply their own buffer for the credentials.

[1]: sock_dtls is a wrapper around libraries that implements DTLS e.g. tinyDTLS, wolfSSL.

pokgak commented:
@miri64  @jcarrano I updated my comment and the PR description. I hope it is easier to understand now. Also fixed requested changes.

As for constfs, credman may need to handle dynamically added credentials during runtime if something like [1] is implemented in the future. So I think it is better to not make the assumption that the credentials is known at compile-time.

[1]: https://tools.ietf.org/html/draft-ietf-ace-dtls-authorize-08

pokgak commented:
Rebased to current master.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,167 @@
+        }
+    }
+    return (used == 0) ? 0 : -1;
+}
```
Can be simplified to

```C
static int _find_next_free_pos(void) {
    int i = _find_credential_pos(CREDMAN_TYPE_EMPTY, CREDMAN_TAG_EMPTY);
    if (i < 0) {
        return (used == 0) ? 0 : -1;
    }
    else {
        return i;
    }
}
```

It also reduces code duplication.

miri64 replied:
Also, the `(used == 0) ? 0 : -1` check is redundant, since `(used == 0)` can only be true if all entries are empty and thus also `credentials[0]`.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,167 @@
+#ifdef TEST_SUITES
+void credman_reset(void)
+{
+    memset(credentials, 0,
```
This function should also be mutex locked.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,167 @@
+    if (pos < 0) {
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
+        ret = CREDMAN_NOT_FOUND;
```
IMHO it is okay to return `CREDMAN_OK` in this case

miri64 replied:
So the whole function can be simplified to

```C
int credman_delete(credman_tag_t tag, credman_type_t type)
{
    mutex_lock(&_mutex);
    int ret = CREDMAN_ERROR;
    int pos = _find_credential_pos(tag, type);
    if (pos >= 0) {
        memset(&credentials[pos], 0, sizeof(credman_credential_t));
        used--;
    }
    mutex_unlock(&_mutex);
    return CREDMAN_OK;
}

pokgak replied:
Going on that thought, I think it is better to also change the return value of `credman_delete` to `void` as returning `CREDMAN_OK` every time is not that useful after all. What's your opinion about this?

miri64 replied:
:+1:

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,167 @@
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
+        ret = CREDMAN_NOT_FOUND;
+        goto end;
```
Goto can be prevented with else

```C
    /* ... */
    if (pos < 0) {
        DEBUG("credman: credential with tag %d and type %d not found\n",
               tag, type);
        ret = CREDMAN_NOT_FOUND;
    }
    else {
        memcpy(credential, &credentials[pos], sizeof(credman_credential_t));
        ret = CREDMAN_OK;
    }
    mutex_unlock(&_mutex);
    return ret;
```

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,185 @@
+} ecdsa_params_t;
+
+/**
+ * @brief Tag of the credential. Must be bigger than 0.
```
This is just the type definition, so adding value restrictions to it seems off (also its "[…] greater than 0").

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,247 @@
+#define CREDMAN_TEST_TAG (1)
+
+static int _compare_credentials(const credman_credential_t *a,
+                               const credman_credential_t *b)
```
Indentation

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,247 @@
+    // delete starting from first added credential
+    TEST_ASSERT_EQUAL_INT(CREDMAN_OK, credman_delete(tag1, in_credential.type));
+    TEST_ASSERT_EQUAL_INT(CREDMAN_OK, credman_delete(tag2, in_credential.type));
+    TEST_ASSERT_EQUAL_INT(0, credman_get_used_count());
```
Please also add a test that tries adding stuff after space what made through deletion

miri64 commented on tests/unittests/tests-credman/tests-credman.h:
```diff
@@ -0,0 +1,44 @@
+#endif
+
+#endif /* TESTS_CREDMAN_H */
+/** @} */
```
Github notes some whitespace error "No newline at end of file" here (also in some of the other files). Please check.

pokgak commented:
Thanks for the review @miri64. I pushed some commits addressing the review comments.

miri64 commented:
Had a look at your changes. Code design looks good now.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,157 @@
+        DEBUG("credman: no space for new credential\n");
+        ret = CREDMAN_NO_SPACE;
+        goto end;
+    }
```
Though I'm wondering if this

```C
pos = _find_credential_pos(tag, type);
if (check pos condition) {
    throw error;
}
```

duplication might be consolidatable somehow. I had no idea yet though.

miri64 replied:
Usually @cgundogan has some nice ideas when it comes to this kind of stuff. Maybe you can ask him :-).

pokgak replied:
This is what I came up with based on suggestion from Cenk:

```
/* Find position of credential with given tag and type.
 * To avoid going through the list twice, this function also returns the
 * first empty position found.
 */
static int _find_credential_pos(credman_tag_t tag, credman_type_t type, int8_t *empty_pos)
{
    *empty_pos = -1;
    for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
        credman_credential_t *c = &credentials[i];
        if ((c->tag == tag) && (c->type == type)) {
            return i;
        }
        /* only check until empty position found */
        if ((empty_pos) && (*empty_pos < 0) &&
            (c->tag == CREDMAN_TAG_EMPTY) && (c->type == CREDMAN_TYPE_EMPTY)) {
            *empty_pos = i;
        }
    }
    return -1;
}

...

/* check if credential already exist and find empty position simultaneously */
int8_t empty_pos;
    pos = _find_credential_pos(credential->tag, credential->type, &empty_pos);
    if (pos >= 0) {
        DEBUG("credman: credential with tag %d and type %d already exist\n",
              credential->tag, credential->type);
        ret = CREDMAN_EXIST;
        goto end;
    }
    if (empty_pos < 0) {
        DEBUG("credman: no space for new credential\n");
        ret = CREDMAN_NO_SPACE;
        goto end;
    }
```

miri64 replied:
I can't find your comment on this in Github, but I found the following in my mails

> This is what I came up with based on suggestion from Cenk:
> ```C
> /* Find position of credential with given tag and type.
>  * To avoid going through the list twice, this function also returns the
>  * first empty position found.
>  */
> static int _find_credential_pos(credman_tag_t tag, credman_type_t type, int8_t *empty_pos)
> {
>     *empty_pos = -1;
>     for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
>         credman_credential_t *c = &credentials[i];
>         if ((c->tag == tag) && (c->type == type)) {
>             return i;
>         }
>         /* only check until empty position found */
>         if ((empty_pos) && (*empty_pos < 0) &&
>             (c->tag == CREDMAN_TAG_EMPTY) && (c->type == CREDMAN_TYPE_EMPTY)) {
>             *empty_pos = i;
>         }
>     }
>     return -1;
> }
> 
> ...
> 
> /* check if credential already exist and find empty position simultaneously */
> int8_t empty_pos;
>     pos = _find_credential_pos(credential->tag, credential->type, &empty_pos);
>     if (pos >= 0) {
>         DEBUG("credman: credential with tag %d and type %d already exist\n",
>               credential->tag, credential->type);
>         ret = CREDMAN_EXIST;
>         goto end;
>     }
>     if (empty_pos < 0) {
>         DEBUG("credman: no space for new credential\n");
>         ret = CREDMAN_NO_SPACE;
>         goto end;
>     }
> ```

How about the following simplification (also note, that I was able to do all without `goto`; I kept the label for the error handling above though):

```C
/* Find position of credential with given tag and type.
 * To avoid going through the list twice, this function also returns the
 * first empty credential pool entry found.
 */
static int _find_credential_pos(credman_tag_t tag, credman_type_t type,
credman_credential_t **empty)
{
    for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
        credman_credential_t *c = &credentials[i];
        if ((c->tag == tag) && (c->type == type)) {
            return i;
        }
        /* only check until empty position found */
        if ((empty) && (*empty != NULL) &&
            (c->tag == CREDMAN_TAG_EMPTY) && (c->type == CREDMAN_TYPE_EMPTY)) {
            *empty = c;
        }
    }
    return -1;
}

...

int credman_add(const credman_credential_t *credential)
{
    credman_credential_t *entry = NULL;
    assert(credential);
...
    pos = _find_credential_pos(credential->tag, credential->type, &entry);
    if (pos >= 0) {
        DEBUG("credman: credential with tag %d and type %d already exist\n",
              credential->tag, credential->type);
        ret = CREDMAN_EXIST;
    }
    else if (entry == NULL) {
        DEBUG("credman: no space for new credential\n");
        ret = CREDMAN_NO_SPACE;
    }
    else {
        *entry = *credential;
        used++;
        ret = CREDMAN_OK;
    }
end:
    mutex_unlock(&_mutex);
    return ret;
}

pokgak replied:
> I can't find your comment on this in Github,

I accidentally deleted my comment while editing it. Sorry.

The code looks better and simpler now even with no `goto`. I'll keep this pattern in mind for next time. I'll push a commit for this later today.

miri64 commented:
`Reviewed: 2-code-design` holds after latest change.

miri64 commented:
I ran the unittests for

- `native` (`tests-credman` and `all`) on a quite Ubuntu 19.04.
```
Operating System Environment
-----------------------------
       Operating System: "Ubuntu" "19.04 (Disco Dingo)"
                 Kernel: Linux 5.0.0-20-generic x86_64 x86_64

Installed compiler toolchains
-----------------------------
             native gcc: gcc (Ubuntu 8.3.0-6ubuntu1) 8.3.0
      arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 8-2018-q4-major) 8.2.1 20181213 (release) [gcc-8-branch revision 267074]
                avr-gcc: avr-gcc (GCC) 5.4.0
       mips-mti-elf-gcc: missing
             msp430-gcc: msp430-gcc (GCC) 4.6.3 20120301 (mspgcc LTS 20120406 unpatched)
   riscv-none-embed-gcc: missing
   xtensa-esp32-elf-gcc: missing
   xtensa-lx106-elf-gcc: missing
                  clang: clang version 8.0.0-3 (tags/RELEASE_800/final)

Installed compiler libs
-----------------------
   arm-none-eabi-newlib: "3.0.0"
    mips-mti-elf-newlib: missing
riscv-none-embed-newlib: missing
xtensa-esp32-elf-newlib: missing
xtensa-lx106-elf-newlib: missing
               avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                  cmake: cmake version 3.13.4
               cppcheck: missing
                doxygen: 1.8.13
                 flake8: 3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.7.3 on Linux
                    git: git version 2.20.1
                   make: GNU Make 4.2.1
                openocd: Open On-Chip Debugger 0.10.0+dev-00910-g4dbcb1e7 (2019-06-17-16:24)
                 python: Python 2.7.16
                python2: Python 2.7.16
                python3: Python 3.7.3
             coccinelle: missing
```
- `iotlab-m3` (`tests-credman` and `all`, for the latter I had to merge in the current master, see #11855)
- `samr21-xpro (only `tests-credman`, `all` does not fit)
- `pba-d-01-kw2x` (only `tests-credman`, `all` does not fit)
- `nrf52dk` (`tests-credman` and `all`)
- `msba2` (`tests-credman` and `all`, for the latter I had to merge in the current master, see #11855; still some [unrelated] tests fail, I confirmed that they also fail on master, see https://github.com/RIOT-OS/RIOT/issues/11861)
```
2019-07-18 12:15:10,577 - INFO # main(): This is RIOT! (Version: 2019.10-devel-91-gaa463-HEAD)
2019-07-18 12:15:18,423 - INFO # .................................................................................................................................................................................................................................................................
2019-07-18 12:15:18,427 - INFO # scanf_float_tests.test_f (tests/unittests/tests-scanf_float/tests-scanf_float.c 37) exp 0 was 1
2019-07-18 12:15:18,427 - INFO # .
2019-07-18 12:15:18,431 - INFO # scanf_float_tests.test_e (tests/unittests/tests-scanf_float/tests-scanf_float.c 38) exp 0 was 1
2019-07-18 12:15:18,431 - INFO # .
2019-07-18 12:15:18,458 - INFO # scanf_float_tests.test_E (tests/unittests/tests-scanf_float/tests-scanf_float.c 39) exp 0 was 1
2019-07-18 12:15:18,459 - INFO # .
2019-07-18 12:15:18,461 - INFO # scanf_float_tests.test_g (tests/unittests/tests-scanf_float/tests-scanf_float.c 40) exp 0 was 1
2019-07-18 12:15:18,461 - INFO # .
2019-07-18 12:15:18,463 - INFO # scanf_float_tests.test_sign (tests/unittests/tests-scanf_float/tests-scanf_float.c 45) exp 0 was 1
2019-07-18 12:15:18,463 - INFO # .
2019-07-18 12:15:18,465 - INFO # scanf_float_tests.test_scientific (tests/unittests/tests-scanf_float/tests-scanf_float.c 53) exp 0 was 1
```
- `z1` (only `tests-credman`, `all` does not fit)
- `arduino-mega2560` (only `tests-credman`, `all` does not fit)


All tests except the noted one on `msba2` ran successful.

For `msba2`, `nrf52dk`, and `arduino-mega2560` I used the nodes connected to https://ci-ilab.imp.fu-berlin.de, everything else I tested locally.

pokgak commented:
Rebased and squashed.

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+ * @file
+ * @brief       (D)TLS credentials management module definitions
+ *
+ * @note        This module DOES NOT copies the credentials into the system. It
```
Grammar
```suggestion
 * @note        This module DOES NOT copy the credentials into the system. It
```

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+ */
+typedef struct {
+    void *s;                /**< Pointer to the buffer */
+    size_t len;             /**< Length of s */
```
```suggestion
    size_t len;             /**< Length of credman_buffer_t::s */
```

This way it can be parsed properly by `doxygen`

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+ */
+typedef struct {
+    credman_buffer_t key;   /**< Key buffer */
+    credman_buffer_t id;    /**< Id buffer */
```
```suggestion
    credman_buffer_t id;    /**< ID buffer */
```

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+ * @brief ECDSA parameters
+ */
+typedef struct {
+    const void * private_key;           /**< Pointer to the private key */
```
```suggestion
    const void *private_key;            /**< Pointer to the private key */
```

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,149 @@
+    return ret;
+}
+
+int credman_get(credman_credential_t *credential,credman_tag_t tag,
```
```suggestion
int credman_get(credman_credential_t *credential, credman_tag_t tag,
```

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -0,0 +1,149 @@
+    int pos = _find_credential_pos(tag, type);
+    if (pos < 0) {
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
```
```suggestion
              tag, type);
```

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+* This file is subject to the terms and conditions of the GNU Lesser
+* General Public License v2.1. See the file LICENSE in the top level
+* directory for more details.
+*/
```
Indentation of comment of (` ` missing at beginning of `*` lines).

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+
+    /* get non-existing credential */
+    ret = credman_get(&out_credential, in_credential.tag,
+                                 in_credential.type);
```
```suggestion
                      in_credential.type);
```

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+    TEST_ASSERT_EQUAL_INT(CREDMAN_OK, ret);
+
+    ret = credman_get(&out_credential, in_credential.tag,
+                                 in_credential.type);
```
Dito

miri64 commented on tests/unittests/tests-credman/tests-credman.h:
```diff
@@ -0,0 +1,44 @@
+
+/**
+*  @brief   The entry point of this test suite.
+*/
```
` ` missing at beginning of `*` lines

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+typedef uint16_t credman_tag_t;
+
+/**
+ * @brief Used to identify credentials for application libraries.
```
I'm not really sure what this is supposed to mean. Maybe give an example in the details?

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+} credman_type_t;
+
+/**
+ * @brief Credential informations
```
```suggestion
 * @brief Credential information
```

(the plural of "information" is "information")

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+ *         credman_credential_t::tag with the value of CREDMAN_TAG_EMPTY OR
+ *         credman_credential_t::type with the value of CREDMAN_TYPE_EMPTY OR
+ *         credman_credential_t::params with invalid credential parameters i.e.
+ *         the key points to NULL or has a length of 0.
```
Instead of `OR` after every line I think it is more readable (especially in rendered doc) when every or-case is documented in its own `@return CREDMAN_INVALID`.

miri64 commented on sys/include/net/credman.h:
```diff
@@ -0,0 +1,192 @@
+                credman_type_t type);
+
+/**
+ * @brief Delete a credential from the system. Does nothing if
```
```suggestion
 * @brief Delete a credential from credential pool. Does nothing if
```

At least if I interpret your comment in l.19 correctly ;-).

miri64 commented on tests/unittests/tests-credman/credentials.h:
```diff
@@ -0,0 +1,58 @@
+ * @{
+ *
+ * @file
+ * @brief       credman test application (PSK and ECC keys)
```
This file does not contain an application, just the `credential` (just move the line below here).

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+    TEST_ASSERT_EQUAL_INT(CREDMAN_INVALID, ret);
+    TEST_ASSERT_EQUAL_INT(exp_count, credman_get_used_count());
+
+    /* fill the system credential buffer */
```
~~system~~?

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+        TEST_ASSERT_EQUAL_INT(++exp_count, credman_get_used_count());
+    }
+
+    /* add to full system credential buffer */
```
~~system~~?

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+    TEST_ASSERT_EQUAL_INT(CREDMAN_OK, ret);
+    TEST_ASSERT_EQUAL_INT(++exp_count, credman_get_used_count());
+
+    /* delete a credential from system buffer */
```
~~system~~?

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+    };
+    TEST_ASSERT_EQUAL_INT(0, credman_get_used_count());
+
+    // fill the system pool, assume CREDMAN_MAX_CREDENTIALS is 2
```
Please use only C-style `/* ... */` comments.

miri64 replied:
Also ~~system~~?

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+    TEST_ASSERT_EQUAL_INT(CREDMAN_OK, credman_add(&in_credential));
+    TEST_ASSERT_EQUAL_INT(2, credman_get_used_count());
+
+    // delete the first credential
```
Dito

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+    credman_delete(tag1, in_credential.type);
+    TEST_ASSERT_EQUAL_INT(1, credman_get_used_count());
+
+    // get the second credential
```
Dito

miri64 commented on tests/unittests/tests-credman/tests-credman.c:
```diff
@@ -0,0 +1,254 @@
+        },
+    };
+
+    // add credentials
```
Here as well, not pointing further occurences

miri64 commented:
Only https://github.com/RIOT-OS/RIOT/pull/11564#discussion_r305236025 left.

pokgak commented:
@miri64 if you noticed, I removed `CREDMAN_TAG_GLOBAL` in this 2a0244a. This is my original idea to use with the second use case described in in this PR:

> 2) the user does not have access to the sock because it is located in a lower leveled library used by user application (e.g. gcoap + DTLS).

But while trying to explain it, I noticed that doing it that way may not be scalable if there is other application like gcoap using it in the future. Basically, my idea was a list of reserved `credman_tag_t` containing the tag used in each application. I already have a new idea how to implement it but maybe it is better to discuss it in a new PR as this PR is full of comments already.

That means, this PR (+ sock_dtls PR which is coming) will only enable support for the first use case:

>  1) the user has direct access to the DTLS sock and can actually give the pointer of the credential buffer to the sock

If you're okay with that, I'll proceed with rebasing and squashing the commits.

miri64 commented on sys/net/credman/credman.c:
```diff
@@ -66,24 +68,22 @@ int credman_add(const credman_credential_t *credential)
-    used++;
-    ret = CREDMAN_OK;
+    else {
+        memcpy(entry, credential, sizeof(credman_credential_t));
```
Side-note: `*entry = *credential` does the same (and should have the same ROM-size) but is type-safe, so I would prefer it like this. But this is not blocking.

miri64 commented:
> But while trying to explain it, I noticed that doing it that way may not be scalable if there is other application like gcoap using it in the future.

Good. Another reason why doc is important. Trying to explain stuff oftentimes helps me rethink stuff :-).

> If you're okay with that, I'll proceed with rebasing and squashing the commits.

Yepp. In case it is needed later it can always be re-added.

I give this a final round of tests, and then I'll ACK. You may squash in the meantime.

pokgak commented:
Rebased and squashed.

miri64 commented:
ACK holds

pokgak commented:
Just noticed I messed up the commit a bit. Fixed and re-pushed the commits.

miri64 commented:
I noticed it already when I ACK'd :-)

miri64 commented:
And go!

benpicco commented:
I'm trying to understand the idea behind the credman tag.
Shouldn't credentials be tied to a specific server? So do we need a separate database to map credman IDs to IP addresses / hostnames?
